sifive-blocks/src/drivers/sifive_ccache0.c

```diff
@@ -5,28 +5,55 @@
 
 #ifdef METAL_SIFIVE_CCACHE0
 
+#include <assert.h>
 #include <metal/drivers/sifive_ccache0.h>
 #include <metal/init.h>
 #include <metal/io.h>
+#include <stddef.h>
 
 #define L2_CONFIG_WAYS_SHIFT 8
 #define L2_CONFIG_WAYS_MASK (0xFF << L2_CONFIG_WAYS_SHIFT)
 
 #define L2_REGW(offset)                                                        \
     __METAL_ACCESS_ONCE(                                                       \
         (__metal_io_u32 *)(METAL_SIFIVE_CCACHE0_0_BASE_ADDRESS + (offset)))
+#define L2_REGD(offset)                                                        \
+    __METAL_ACCESS_ONCE(                                                       \
+        (volatile uint64_t *)(METAL_SIFIVE_CCACHE0_0_BASE_ADDRESS + (offset)))
+
+/* Linker symbols to calculate LIM allocated size */
+extern char metal_segment_lim_target_start, metal_segment_lim_target_end;
 
 void sifive_ccache0_init(uint32_t ways) {
     sifive_ccache0_set_enabled_ways(ways);
 }
 
 METAL_CONSTRUCTOR(init_sifive_ccache0) {
-    /* Get the number of available ways per bank */
-    const uint32_t ways = __METAL_GET_FIELD(
+    size_t lim_size = 0;
```

Comment:
This appears to assume that the only lim area used by the application is for code? I'd think we'd want to allow code, data and bss to all land in the LIM. Is that even possible? Would we have to initialize the cache controller earlier (like before any data/bss use?)

Reply:
I'm going to leave additional ways to use the LIM as a future todo, since this brings it in-line with v20.08.